### DISCUSSION TITLE
Sending detection notification via SMS using waveshare 7600 4g

### DISCUSSION THREAD

**Discussion by armsto88** [mcguirepr89/BirdNET-Pi#890] [Q&A]
Hi all,

I would like to receive an sms when a detection has been made. I am tracking a particular bird and plan to install the units in a remote location. I have a waveshare 7600 4g HAT. Looking for guidance on how to make it work with this HAT or other suggestions to archive the same result. 
The voluntary project will be monitoring the threated glossy-black cockatoo, native to Eastern Australia. 
Appreciate any assistance. 
Cheers 
Tom Armstrong

**Comment by ampsu** [comment]
I guess the easiest way is to use built-in Apprise integration which has many SMS services available: https://github.com/caronc/apprise/wiki (Twilio for example). In order to send notifications natively with SMS, modifications to the code are needed by someone who knows how do them (I dont). If you have cellular signal, it is also possible to use other services like Slack etc.

I am personally using Slack+JSON POST to my own databases.

**Reply by WanderingStar** [reply]
A downside to using Apprise to contact an SMS service is that it requires an internet connection. I would expect that will use a lot more cellular data than sending an SMS directly from the hardware.

As you point out, though, the upside is that it's easy.

**Reply by ampsu** [reply]
Sure. It requires a 3G/4G connection but it does not use much data. I am sending out like 800 notifications per day and I also remotely listen to recordings. Data usage is like 0.8-1GB/month (mostly due to sound files) so it translates to like 1€ in terms of cost. I think one Apprise request is like 300bytes or so. It also depends if you want to send every detection or only certain species. I would not send every detection with a SMS, imagine that messaging inbox :D

For eg. for Slack I created different bots which filter and sort detections to species specific channels so it is easier to scroll through daily and all the big data goes to MongoDB. Bear mind that if it is a scientific project, good data collection is a must. I dont know how you can later export and sort all the data from those SMS messages.

**Reply by WanderingStar** [reply]
Yeah, I'd go a step further and suggest that notifications aren't a good mechanism for exporting data for comprehensive analysis. It's easy for notifications to get dropped.

**Reply by ampsu** [reply]
Yeah, sometimes cell tower could be down and your data wont arrive or server sided issues etc. Mission-critical data must always be mirrored in a few places, local db serves some purpose for that. But regarding reliability. I got two Pi-s that only listen to owls. Slack notifications and POSTs match exactly. I havent really compared them to a local db yet but I guess they are close enough. But well hey, things are never 100% reliable. Starting from false detects and ending with some misc system issues. That is why results should be anyways double-checked by human ear.

**Reply by armsto88** [reply]
Thank you for the responses! Appreciate it. Excuse my ignorance , as this is my first project and have limited experience with anything Pi. @ampsu would you mind outlining your set up? How are you getting your 4g ? And how is that working with your slack? I am indifference to whether the notifications are  SMS or any other means. Having access to the complete data including reference audio would be nice but priority is be notified close to when the call happened so i can track bird over a potential flight path with multiple units.
Cheers Tom

**Comment by WanderingStar** [comment]
Hi Tom. I have no experience with that particular HAT (or indeed with cellular HATs in general), but I do have some thoughts about the problem. A few questions:
1. How timely do notifications need to be? If a notification is sent within an hour, is that good enough? Within a day?
2. Do you want to be notified every time there's a detection, or is it acceptable to group the detections? I ask this because I was just camping with my BirdNet installation and got 1-200 detections an hour, which would be a lot of SMSes. Birdnet's notification options provide some controls about how to group and limit the notifications, but they may not be specific enough for your needs. For example, you can exclude common species, but it sounds like you might want to only _include_ one species.
4. Is it important that it's literally an SMS? Is that for cost/power reasons? Will the SIM card in the HAT have internet data capability or just SMS? As @ampsu suggests above, you might be able to use an internet service with Apprise with very low effort.

One thing you could do is write a script to look at the extracted data on disk and run that script from a cron job. You could use something like

```
> find /home/pi/BirdSongs/ -name 'Painted_Bunting*.png' # add some time limits...
...
/home/pi/BirdSongs/Extracted/By_Date/2023-05-02/Painted_Bunting/Painted_Bunting-77-2023-05-02-birdnet-06:33:24.mp3.png
/home/pi/BirdSongs/Extracted/By_Date/2023-05-02/Painted_Bunting/Painted_Bunting-89-2023-05-02-birdnet-06:41:54.mp3.png
/home/pi/BirdSongs/Extracted/By_Date/2023-05-02/Painted_Bunting/Painted_Bunting-86-2023-05-02-birdnet-07:42:48.mp3.png
```

And process that into the text of an SMS, which you then send.

**Reply by WanderingStar** [reply]
Or if it doesn't need to be a literal SMS, your cron job could POST that data to a URL (if you don't want to set up a server to collect this, maybe posting to a Google Form to store the data in a Google Sheet would work), email you a daily CSV file, directly connect to a database you control and store the data, or a hundred other things, depending on what's convenient for you.

**Reply by ehpersonal38** [reply]
> For example, you can exclude common species, but it sounds like you might want to only include one species.

FYI In a recent update this functionality was also added. https://github.com/mcguirepr89/BirdNET-Pi/commit/6d3857b86e5f042ea1dbf3d8a55baaedf31059e1

**Reply by armsto88** [reply]
Thanks for your response. 
1. As Timely as possible or a time stamped daily csv.
2. I will have it set to just the Glossy black cockatoo for this project which is thretened so have spent a good deal of time looking for them!!
3. Not important to be an SMS

Very interested in this Cron Job business and producing google sheets with daily data. That sounds like knowledge that would cover me for future projects ! POST to a URL ? How do i learn more about this?  Sorry for being a noob.  I will be trying to actively work it out as i wait for a response ! cheers Tom

**Reply by WanderingStar** [reply]
Have you used that Waveshare HAT before? [This](https://www.waveshare.com/wiki/SIM7600E-H_4G_HAT#Raspbian) makes it look like you can get it to act as a normal network adapter by plugging its USB port into a USB port on the RPi. Then you can use Apprise as @ampsu suggests above and choose any [notification service](https://github.com/caronc/apprise/wiki) that's convenient for you. You can probably use a [Form](https://github.com/caronc/apprise/wiki/Notify_Custom_Form) notification to [post to a Google Form](https://yaz.in/p/submitting-a-google-form-using-the-command-line/).

It also seems like you can access it over the RPi's serial interface,and send SMS messages. That might be better if you're planning to put it in an enclosure that doesn't make it easy to run a USB cable around the side. In that case, you probably don't want to use Apprise, but instead run a script that formulates the SMS and then sends it using the serial AT commands in the Waveshare docs.

Maybe something like:
```
find /home/pi/BirdSongs/ -name 'Painted_Bunting*.png' -cmin -5 | \
  perl -pe 's{^.*/(\w+)-(\d+)-(\d\d\d\d-\d\d-\d\d)-birdnet-(\d\d:\d\d:\d\d)\.mp3\.png}{$1 seen with confidence $2 at $3 $4}' > sms.txt
```
And then send the appropriate AT commands to `/dev/ttyS0` to send an SMS, and then send sms.txt to that dev.

Set that in a cron job to run every 5 minutes (or whatever interval you want, if you change the `-cmin`).

**Reply by armsto88** [reply]
I have never used a pi before ! All very new and slightly daunting. I have been looking at ways to do daily syncs of the "extracted" folder to a cloud. This in conjuncition with some form of detection push notification would be ideal. I have been looking into something called Rclone for cloud syncs as well as trying to work out  "slack" for notifications. Any experience with rclone ? Or anything you have used for cloud syncing ?

**Reply by WanderingStar** [reply]
Aha. Do you have experience with Linux/Unix command lines? I use `rsync` to transfer files daily to another server that I run. I haven't used `rclone`, but `rclone copy` looks like it would work, if you have a functional network connection.

**Reply by armsto88** [reply]
I have a bit more since we have began this chat ! Prior to that none. I have got it all running with the rsync transfering files to a google drive. Now just trying to get the pi connect to the 4g output of the HAT.

**Comment by Mattmanandeddie** [comment]
Would something like this work for you?

Sending emails to SMS or MMS
AT&T: [number@txt.att.net](mailto:number@txt.att.net) (SMS), [number@mms.att.net](mailto:number@mms.att.net) (MMS)
T-Mobile: [number@tmomail.net](mailto:number@tmomail.net) (SMS & MMS)
Verizon: [number@vtext.com](mailto:number@vtext.com) (SMS), [number@vzwpix.com](mailto:number@vzwpix.com) (MMS)
Sprint: [number@messaging.sprintpcs.com](mailto:number@messaging.sprintpcs.com) (SMS), [number@pm.sprint.com](mailto:number@pm.sprint.com) (MMS)
